text/0000-raw-reference-operator.md

```diff
@@ -0,0 +1,189 @@
+let _ = &X; // throwing away the value immediately changes nothing
+&X; // different syntax for the same thing
+
+let x = &X as &T as *const T; // this is casting to raw "too late"
+```
+
+The only way to create a pointer to an unaligned or dangling location without
+triggering undefined behavior is to *immediately* turn it into a raw pointer.
+All of the following are valid:
+
+```rust
+let packed_cast = unsafe { &packed.field as *const _ };
+let packed_coercion: *const _ = unsafe { &packed.field };
+let null_cast: *const _ = unsafe { &*ptr::null() } as *const _;
+let null_coercion: *const _ = unsafe { &*ptr::null() };
+```
+
+The intention is to cover all cases where a reference, just created, is
+immediately explicitly used as a value of raw pointer type.
+
+These two operations (taking a reference, casting/coercing to a raw pointer) are
+actually considered a single operation happening in one step, and hence the
+invariants incurred by references do not come into play.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When translating HIR to MIR, we recognize `&[mut] <place> as *[mut|const] ?T`
+(where `?T` can be any type, also a partial one like `_`) as well as coercions
+from `&[mut] <place>` to a raw pointer type as a special pattern and turn them
+into a single MIR `Rvalue` that takes the address and produces it as a raw
+pointer -- a "take raw reference" operation.  This might be a variant of the
+existing `Ref` operation (say, a boolean flag for whether this is raw), or a new
+`Rvalue` variant.  The borrow checker should do the usual checks on `<place>`,
+but can just ignore the result of this operation and the newly created
+"reference" can have any lifetime.  (Currently this will be some form of
+unbounded inference variable because the only use is a cast-to-raw, the new "raw
+reference" operation can have the same behavior.)  When translating MIR to LLVM,
+nothing special has to happen as references and raw pointers have the same LLVM
+type anyway; the new operation behaves like `Ref`.
+
+When interpreting MIR in the miri engine, the engine will recognize that the
+value produced by this `Rvalue` has raw pointer type, and hence must not satisfy
+any special invariants.
+
+When doing unsafety checking, we make references to packed fields that do *not*
+use this new "raw reference" operation a *hard error even in unsafe blocks*
+(after a transition period).  There is no situation in which this code is okay;
+it creates a reference that violates basic invariants.  Taking a raw reference
+to a packed field, on the other hand, is a safe operation as the raw pointer
+comes with no special promises.  "Unsafety checking" is thus not even a good
+term for this, maybe it should be a special pass dedicated to packed fields
+traversing MIR, or this can happen when lowering HIR to MIR.  This check has
+nothing to do with whether we are in an unsafe block or not.
```

Comment:
This compiles without errors on stable Rust today:

```rust
#[repr(packed)]
pub struct A {
    x: u8,
    y: u64
}

pub fn foo(mut _x: A) {
    let y = &mut _x.y;
    *y = 3;
}
```

but the text does not directly mention whether this will be an error too. It hints that this will be the case when it says (emphasis mine): "hard error **even** in unsafe blocks", but it should probably spell this out.

Reply:
No it doesn't

```
warning: borrow of packed field is unsafe and requires unsafe function or block (error E0133)
 --> src/lib.rs:8:13
  |
8 |     let y = &mut _x.y;
  |             ^^^^^^^^^
  |
  = note: #[warn(safe_packed_borrows)] on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>
  = note: fields of packed structs might be misaligned: dereferencing a misaligned pointer or even just creating a misaligned reference is undefined behavior
```

Reply:
But that's not an error, that's a warning, and it exists since Rust 1.29, while `repr(packed)` is stable since Rust 1.0.

Reply:
Yes, we tell LLVM that references are aligned, so this is already UB and AFAIK has always been.

Reply:
It's a future compat warning, as the text quite clearly says. The only thing blocking us from making it an error is resolving this RFC :)

Reply:
> his RFC does not turn that warning into an error 

This RFC wouldn't make it a compiler error, but it is still UB and will continue to be. In the future, we intend to make it a hard error. It has *always* been the case that taking a reference to a field that is not properly aligned is UB.

Reply:
> This RFC wouldn't make it a compiler error, but it is still UB and will continue to be. 

I don't know why you and @arielb1 keep bringing up that this is UB. Who is saying otherwise ?

> so we cannot soundly allow this in safe code.

This RFC does state that `&mut T as *mut T` is always safe, do you disagree with that statement?

Reply:
Re-reading my own RFC text, the changes around what is and is not permitted after actually part of the normative text of the RFC.  Should I move them out?  I think it makes sense to consider this entire somewhat messy situation together, but I am not sure if the RFC should actually be a decision on all of them.

Reply:
I think it is better to leave this out and to better focus on this change only. 

You can mention that there is an issue open about turning a particular warning into an error, that's currently blocked on having a way to construct a pointer to a field without going through a reference, and that this RFC would solve that blocker. But I wouldn't make the text about that.